src/dcp/dcp.c

```diff
@@ -371,9 +616,62 @@ int main(int argc, char** argv)
     /* free the copy options */
     mfu_walk_opts_delete(&walk_opts);
 
-    /* shut down MPI */
-    /* shut down MPI */
+    /* DAOS: unmount DFS, and close containers and pools */
+#ifdef DAOS_SUPPORT
+    if (mfu_src_file->type == DAOS) {
+        rc = dfs_umount(mfu_src_file->dfs);
+        MPI_Barrier(MPI_COMM_WORLD);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "Failed to mount DFS namespace");
+        }
+        rc = daos_cont_close(src_coh, NULL);
+        MPI_Barrier(MPI_COMM_WORLD);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "Failed to close container (%d)", rc);
+        }
+    }
+
+    if (mfu_dst_file->type == DAOS) {
+        rc = dfs_umount(mfu_dst_file->dfs);
+        MPI_Barrier(MPI_COMM_WORLD);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "Failed unmount DFS namespace");
+        }
+        rc = daos_cont_close(dst_coh, NULL);
+        MPI_Barrier(MPI_COMM_WORLD);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "Failed to close container (%d)", rc);
+        }
+    }
+
+    if (mfu_src_file->type == DAOS) {
+        rc = daos_pool_disconnect(src_poh, NULL);
+        MPI_Barrier(MPI_COMM_WORLD);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "Failed to disconnect from source pool");
+        }
+    }
+
+    if (mfu_dst_file->type == DAOS && !same_pool) {
+        rc = daos_pool_disconnect(dst_poh, NULL);
+        MPI_Barrier(MPI_COMM_WORLD);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "Failed to disconnect from destination pool");
+        }
+    }
+#endif
+
+    /* free the mfu_file object */
+    mfu_file_delete(&mfu_src_file);
+    mfu_file_delete(&mfu_dst_file);
+
+#ifdef DAOS_SUPPORT
+    /* finalize daos */
+    rc = daos_fini();
```

Comment:
Assuming we add the check for DAOS around the section that contains the daos_init, we'll need the same check down here.

Reply:
Nevermind about this, given our earlier discussion about daos_init.